fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java

```diff
@@ -1209,6 +1212,170 @@ public RecordCursor<IndexEntry> scanIndex(@Nonnull Index index, @Nonnull IndexSc
+        IndexMaintainer indexMaintainer = getIndexMaintainer(index);
+        final Tuple hopInfo = createHopInfo(indexMaintainer.getIndexSubspace(), recordSubspace, commonPrimaryKey, index);
+        // Get the cursor with the index entries and the records from FDB
+        RecordCursor<FDBIndexedRawRecord> indexEntries = indexMaintainer.scanIndexPrefetch(range, hopInfo.pack(), continuation, scanProperties);
+        // Parse the index entries and payload and build records
+        RecordCursor<FDBIndexedRecord<M>> indexedRecordCursor = indexEntriesToIndexRecords(scanProperties, orphanBehavior, recordSubspace, indexEntries, typedSerializer);
+
+        return context.instrument(FDBStoreTimer.Events.SCAN_INDEX_KEYS, indexedRecordCursor);
+    }
+
+    @VisibleForTesting
+    @Nonnull
+    <M extends Message> RecordCursor<FDBIndexedRecord<M>> indexEntriesToIndexRecords(@Nonnull final ScanProperties scanProperties,
+                                                                                     @Nonnull final IndexOrphanBehavior orphanBehavior,
+                                                                                     @Nonnull final Subspace recordSubspace,
+                                                                                     @Nonnull final RecordCursor<FDBIndexedRawRecord> indexEntries,
+                                                                                     @Nonnull RecordSerializer<M> typedSerializer) {
+        SplitHelper.SizeInfo sizeInfo = new SplitHelper.SizeInfo();
+        ExecuteState executeState = scanProperties.getExecuteProperties().getState();
+
+        RecordCursor<FDBIndexedRecord<M>> indexedRecordCursor = indexEntries.mapPipelined(indexedRawRecord -> {
+            // Use the raw record entries to reconstruct the original raw record (include all splits and version, if applicable)
+            FDBRawRecord fdbRawRecord = reconstructSingleRecord(recordSubspace, sizeInfo, indexedRawRecord.getRawRecord(), scanProperties, useOldVersionFormat());
+            if (fdbRawRecord == null) {
+                return handleOrphanEntry(indexedRawRecord.getIndexEntry(), orphanBehavior);
+            } else {
+                Optional<CompletableFuture<FDBRecordVersion>> versionFutureOptional = Optional.empty();
+                // The version future will be ignored in case the record already has a version
+                if (useOldVersionFormat() && !fdbRawRecord.hasVersion()) {
+                    versionFutureOptional = loadRecordVersionAsync(indexedRawRecord.getIndexEntry().getPrimaryKey());
+                }
+                final ByteScanLimiter byteScanLimiter = executeState.getByteScanLimiter();
+                // TODO: Should we add the index entry bytes to the scanned bytes count?
+                // TODO: Since the prefetched bytes are already brought in, shouldn't this even be considered here? Stopping the iteration will just ignore bytes that were already read
```

Comment:
Yes, I think the index entry bytes should be included. It is already the case that unread key-value pairs may be pending in an `AsyncIterator` beyond what triggers a byte-scan limit.

Reply:
Done